calico/felix/config.py

```diff
@@ -46,166 +46,299 @@
              "crit":      logging.CRITICAL,
              "critical":  logging.CRITICAL}
 
+# Sources of a configuration parameter. The order is highest-priority first.
+DEFAULT = "Default"
+ENV = "Environment variable"
+FILE = "Configuration file"
+GLOBAL_ETCD = "Global etcd configuration"
+LOCAL_ETCD = "Host specific etcd configuration"
+DEFAULT_SOURCES = [ ENV, FILE, GLOBAL_ETCD, LOCAL_ETCD ]
 
 class ConfigException(Exception):
-    def __init__(self, message, source):
+    def __init__(self, message, parameter):
         super(ConfigException, self).__init__(message)
         self.message = message
-        self.source = source
+        self.parameter = parameter
 
     def __str__(self):
-        return "%s (data source : %s)" % (self.message, self.source)
+        return "%s (value %r for %s (%s), read from %r)" \
+            % (self.message,
+               self.parameter.value,
+               self.parameter.name,
+               self.parameter.description,
+               self.parameter.active_source)
+
+class ConfigParameter(object):
+    """
+    A configuration parameter. This contains the following information.
+    - The name of the field.
+    - Where the location can validly be read from
+    - The current value
+    - Where the value was read from
+    """
+    def __init__(self, name, description, default,
+                 sources=DEFAULT_SOURCES, int=False):
+        """
+        Create a configuration parameter.
+        :param str description: Description for logging
+        :param list sources: List of valid sources to try
+        :param str default: Default value
+        :param bool int: Integer value?
+        """
+        self.description = description
+        self.name = name
+        self.sources = sources
+        self.value = default
+        self.active_source = None
+        self.value_is_int = int
+
+    def set(self, value, source):
+        """
+        Set a value of a parameter - unless already set.
+        :param value: value
+        :param source: source; for example "Configuration file /etc/felix.cfg"
+        """
+        if self.active_source is None:
+            log.debug("Read value %r for %s (%s) from %r",
+                      value,
+                      self.name,
+                      self.description,
+                      source)
+
+            self.active_source = source
+
+            if self.value_is_int:
+                self.value = value
```

Comment:
This double assignment looks weird, but reading the code makes it apparent that it's for the benefit of the `ConfigException`. A comment might be a good idea here.